### ISSUE TITLE
flake8==3.8.1 excute failed because importlib_metadata upgrade

### ISSUE THREAD

**Issue by hanhouchao** [python/importlib_metadata#259]
error message 

```
Traceback (most recent call last):
  File "/usr/local/bin/flake8", line 7, in <module>
    from flake8.main.cli import main
  File "/usr/local/lib/python3.5/dist-packages/flake8/main/cli.py", line 5, in <module>
    from flake8.main import application
  File "/usr/local/lib/python3.5/dist-packages/flake8/main/application.py", line 11, in <module>
    from flake8 import checker
  File "/usr/local/lib/python3.5/dist-packages/flake8/checker.py", line 18, in <module>
    from flake8 import processor
  File "/usr/local/lib/python3.5/dist-packages/flake8/processor.py", line 13, in <module>
    from flake8 import utils
  File "/usr/local/lib/python3.5/dist-packages/flake8/utils.py", line 16, in <module>
    from flake8._compat import lru_cache
  File "/usr/local/lib/python3.5/dist-packages/flake8/_compat.py", line 12, in <module>
    import importlib_metadata
  File "/usr/local/lib/python3.5/dist-packages/importlib_metadata/__init__.py", line 43, in <module>
    class PackageNotFoundError(ModuleNotFoundError):
NameError: name 'ModuleNotFoundError' is not defined
```

**Comment by toabi** [comment]
Yes, facing the same issue where my Ansible setup dies on a Ubuntu 16 Host because it only has Python 3.5…

**Comment by cboulay** [comment]
I believe this is the same error causing install-qt-action to fail: https://github.com/jurplel/install-qt-action/issues/65

**Comment by cboulay** [comment]
Unfortunately, @toabi 's solution didn't work for me (and neither for the install-qt-action maintainer who tried it). With a small amount of investigation, it looks like the offending line was introduced on May 19, 2019 (https://github.com/python/importlib_metadata/commit/73651fd9b8df0c9c79de087b61246294bf89418f) and the last release before then was 0.12.

Ubuntu 16.04 is supported until April 2021, and we'd like our packages to continue to work on then, but the GitHub Actions Ubuntu 16.04 runners come with Python 3.5, so that means our build & test routines need to work on Python 3.5. Unfortunately, `ModuleNotFoundError` was introduced in Python 3.6, so the routine to install Qt, which depends on importlib_metadata for Python < 3.8, fails.

I'm not very familiar with this package. I only use it second-hand through the qt-installer. It's a bit curious that this error only seems to be popping up recently, despite the change being about 1.5 years ago.

Anyway, a potentially simple fix might be to replace `class PackageNotFoundError(ModuleNotFoundError):` with `class PackageNotFoundError(ImportError ):` . However, as I'm not familiar with this package, I don't know what the implications of that are.

**Comment by warsaw** [comment]
Can you use our [official Docker image](https://gitlab.com/python-devs/ci-images/-/tree/master) based on Ubuntu 20.04?

**Comment by cboulay** [comment]
@warsaw 

Sorry I'm a bit confused by that. Are you suggesting to run a Ubuntu 20.04 docker image inside the Ubuntu 16.04 runner in GitHub Actions? I don't know enough about the qt-installer to know if it requires sudo access to the host filesystem (seems likely) and I don't know enough about docker to know if it can grant such access. It seems like a big ask for me to ask the install-qt-action developers to do this.

For clarification, I'm testing and deploying a C++ application, but that C++ application uses Qt, and the helper action[1] to install Qt uses python scripts to do so. Once Qt is installed the rest of the build uses CMake / make / gcc and that MUST happen inside the host Ubuntu 16.04 environment (not docker) to ensure compatibility with users stuck on Ubuntu 16.04.

I know I laid out a very particular use case, but in my searching I've encountered a few other reports elsewhere that lead back to the exact same issue, so a fix in importlib_metadata will likely help others too.

1. The `apt-get install` version of Qt available to Ubuntu 16.04 is Qt 5.5, which is missing some features I'd like. I could code around them if necessary. However, the Qt installer works well on Windows, MacOS, and all other versions of Ubuntu (18.04 and latest==20.04) so I would like to use it to both give me access to newer Qt on Ubuntu 16.04 and to simplify my CI scripts.

**Comment by alemuntoni** [comment]
In my pipeline, I am forced to use ubuntu-16.04 with newer Qt.
It is a common practice on Linux to build and deploy applications in the oldest still-supported distribution, to maximize compatibility.
The tool [linuxdeployqt](https://github.com/probonopd/linuxdeployqt) forces to use the oldest still-supported distribution. It just doesn't work on newer distros. [MeshLab](https://github.com/cnr-isti-vclab/meshlab) relies on this tool for linux, and therefore MeshLab's Linux builds have been broken for 4 days due to this issue.

**Comment by hanhouchao** [comment]
> Yes, facing the same issue where my Ansible setup dies on a Ubuntu 16 Host because it only has Python 3.5…
> 
> Quickfixed by downgrading to `pip3 install importlib-metadata==2.1.0`

临时这么解决的，但是不应该出现这样不兼容的情况的

**Comment by FFY00** [comment]
Sorry, how is `importlib_metadata` getting in your system? Python 3.5 support was dropped in 3.0.0, the latest compatible version is 2.1.0. This behavior is expected, even you fixed the `ModuleNotFoundError` issue you would likely get other issues.

Neither pip nor your distribution package manager should be installing importlib_metadata 3.* on Python 3.5. So, I'd say the issue is there.
If this is code Ansible is trying to run on the target system, which has Python 3.5, then it is an Ansible issue. Ansible should check for the remote Python version and see if it is supported. I am not familiar on how Ansible runs code on targets but it would seem pretty weird to be that it'd be bringing its own importlib_metadata.

TLDR: This importlib_metadata version should have never gotten into the system as it does not support the Python version there. You need an older version which still has support for Python 3.5.

**Comment by cboulay** [comment]
Hi @FFY00 , thank you for pointing me in the right direction.

The GitHub Actions Ubuntu 16.04 runner comes with pip 8 which doesn't respect the python_requires tag. A pip update before attempting to install helped me get on my way. For me at least, my issue lies with install-qt-action and I have a workaround [described here](https://github.com/jurplel/install-qt-action/issues/65#issuecomment-735513839).

**Comment by FFY00** [comment]
Oh, yes. That pip version is very old, probably before `python_requires` was a thing.

**Comment by toabi** [comment]
Indeed the issue arises when using an outdated pip version. Upgrading it solves the issue.

**Comment by jaraco** [comment]
Thanks FFY00 for identifying the issue. There's nothing importlib_metadata can do here. Users requiring support for older versions of pip will need to pin their dependencies (as we all did back in those days). I'll pin this issue for a while to increase visibility.
